design-docs/007-basic-expiration.md

```diff
@@ -0,0 +1,561 @@
+the "Update `A` in place" case, because all snapshot ids have
+been preserved.
+
+In terms of concurrency safety, both algorithms have similar properties.
+We  don't evaluate the case of multiple expiration operations ran at
+once. We keep undefined behavior in that case.
+
+If a user is reading a snapshot being rewritten by the left algorithm,
+they could either get the history version before or after the operation,
+but never an invalid view of history or an invalid snapshot. Object stores
+guarantee atomic updates to individual objects, and they also guarantee
+strong read-after-update consistency (with some region caveats in the case of
+Tigris).
+
+For the right algorithm case, the "race condition" is produced when the
+branch is updated, but again, this is safe and atomic.
+
+Any operations that may require multiple passes through the repository
+history, will suffer inconsistencies if expiration is running concurrently.
+They can get different views of the history in each pass.
+But that's true for any algorithm that modifies
+history. Users will need to implement their own consistency mechanisms
+for this operations, and to avoid multiple expiration operations
+happening at once. Expiration is considered an administrative operation,
+and in general avoided during repository usage.
+
+We choose to implement the algorithm on the left: "Update A in place" for its
+simplicity. We don't think the alternative has enough advantages to justify
+its shortcomings.
+
+
+## Using this algorithm for snapshot expiration
+
+Now that we know how to abandon snapshots, we can expire snapshots based
+on age. Here is a rough pseudo-code overview of the algorithm:
+
+```python
+for ref in repo.tags + repo.branches:
+    snapshot_id = resolve_ref(ref)
+    snapshot = fetch_snapshot(snapshot_id)
+
+    # we do nothing if the ref points to a branch that is fully expired already
+    if snapshot.timestamp < TIME_THRESHOLD:
+        continue
+
+    # find the root of the tree
+    # this is B in our algorithm, the root
+    root = next(for parent in snapshot.ancestry if parent.parent is None)
+
+    # find the last non-expired snapshot
+    # this is A in our algorithm
+    editable = last(for parent in snapshot.ancestry if parent.timestamp >= TIME_THRESHOLD)
+    editable.parent = root
+    write_snapshot(editable)
+```
+
+This process can be executed with branches and tags in any order. In the end,
+the expired snapshots will be left with either, no branches/tags pointing to them,
+or with branches or tags pointed directly to them, with no intermediate non-expired snapshots
+such as `N1` and `N2`.
+
+This end state is the ideal one for the garbage collection algorithm to start.
+Non-pointed snapshots will be deleted, together with their exclusive manifests
+and chunks. Branches or tags pointing directly to expired snapshots can be
+deleted too, which will generate extra garbage GC can clean.
+
+### Sample run of the algorithm:
+
```

Comment:
Before going to this intricate, complicated example, it might be nice to see a much simpler scenario with a single branch and a linear history, where we just want to expire everything older than a certain timedelta.